### ISSUE TITLE
Evaluate metrics during training is so much different after dataset fully predicted

### ISSUE THREAD

**Issue by ysssgdhr** [MIC-DKFZ/nnUNet#404]
Hello Fabian!
My current work is unsupervised domain adaptation of kidney and tumor in abdominal CT. In order to see the effect of training, I modified the code so that the validation set during training is the test set (the validation set is the preprocessed test set, using exactly the same z-score and plans.pkl file).
As shown in progress.png, the evaluation metrics are finally close to 0.7.
![image](https://user-images.githubusercontent.com/24486813/100309332-9b08e600-2fe5-11eb-9232-371a8cffdec3.png)

After training, I used the nnUnet_predict command to predict the test set. I stripped out the code of evaluate segmentation results in nnUnet, and compared the ground truth and prediction results. It is found that the dice is very bad, unlike during training, the dice of cancer is only 24%.
![image](https://user-images.githubusercontent.com/24486813/100309386-b70c8780-2fe5-11eb-9ca0-24c6ce2f0010.png)

I don't understand why the metrics can reach 0.7 during training, and the validation set and test set are exactly the same. How can I modify the code to make the metrics in progress.png truly reflect the level of the model?

**Comment by ysssgdhr** [comment]
I read the code and i think global_dc_per_class do not include background , right?
Maybe I am wrong?

**Comment by ysssgdhr** [comment]
[https://github.com/MIC-DKFZ/nnUNet/issues/277#issuecomment-681729991](url)
Sorry Fabian, during the last three months I have been finding job after graduation. It is a campus autumn recruitment peak season. Sorry for delay response again.

Here is the z-score of KITS and OURS:
KITS: 
![image](https://user-images.githubusercontent.com/24486813/100314073-d0ff9780-2ff0-11eb-86c3-b3f839d9a2ac.png)
OURS:
![image](https://user-images.githubusercontent.com/24486813/100314123-e96fb200-2ff0-11eb-9f24-70f5fc4019e0.png)

And info about sizes and spacings:
KITS:
![image](https://user-images.githubusercontent.com/24486813/100314263-381d4c00-2ff1-11eb-8c28-38ab771d7f7b.png)
![image](https://user-images.githubusercontent.com/24486813/100314450-9c401000-2ff1-11eb-9426-cefed873499f.png)
OURS:
![image](https://user-images.githubusercontent.com/24486813/100314279-42d7e100-2ff1-11eb-8d69-01d16a6e8881.png)
![image](https://user-images.githubusercontent.com/24486813/100314524-cb568180-2ff1-11eb-9f12-2be25ce8da46.png)

In the past three months, my work has gradually made new progress. My graduation project is to improve the unsupervised domain adaptation effect of the model training in the direction of OURS->KITS.

First, all CTs of OURS are small patches with a fixed size of (200, 150, 150) divided into independent left and right kidneys. The model trained on OURS cannot learn any information except a small area around the kidney. 
It is foolish to evaluate the effect directly on the KITS dataset like I used to do. Correspondingly, I crop the patch of the corresponding size from KITS according to Ground truth, and make sure that the size of the patch after resampling according to the spacing of [0.6816408, 0.6806643, 0.6816408] is the same as OURS (200, 150, 150).
patch from KITS:
![image](https://user-images.githubusercontent.com/24486813/100314985-cb0ab600-2ff2-11eb-819e-cf78a7f74f1f.png)
![image](https://user-images.githubusercontent.com/24486813/100315071-fab9be00-2ff2-11eb-98a8-efb9e533afb4.png)
patch from OURS:
![image](https://user-images.githubusercontent.com/24486813/100315027-e675c100-2ff2-11eb-820c-529d4dfc1e42.png)
![image](https://user-images.githubusercontent.com/24486813/100315051-ef669280-2ff2-11eb-8123-91e36eea1954.png)

What you said on the previous issue webpage is completely correct. If OURS is a "simple" data set, then KITS is undoubtedly a "difficult" data set. I contacted the official staff of PAIP2019 and learned that KITS's data was collected from more than 50 data centers, while OURS only came from one hospital. Therefore, data homogeneity is the biggest problem. Models cannot learn diverse features and will not learn to cope with resampling artifacts.
OURS:
![image](https://user-images.githubusercontent.com/24486813/100315731-11144980-2ff4-11eb-9b9a-4badfb71e0a2.png)
KITS:
![image](https://user-images.githubusercontent.com/24486813/100315755-1a051b00-2ff4-11eb-8a6f-ba78d4f6e50c.png)

I limit all training methods and prediction methods to 2D Unet,  so that I can use many novel unsupervised domain adaptation methods of CVPR2018 and 2019 to improve the effect of model. (Unsupervised research on 3D images is very scarce)
The plans.pkl file tells me that batch_size should be set to 108, is this normal?
_data_identifier
nnUNetData_plans_v2.1_2D
plans_per_stage
{0: {'batch_size': 108, 'num_pool_per_axis': [5, 5], 'patch_size': array([192, 160]), 'median_patient_size_in_voxels': array([143, 191, 143]), 'current_spacing': array([0.6816408, 0.6806643, 0.6816408]), 'original_spacing': array([0.6816408, 0.6806643, 0.6816408]), 'pool_op_kernel_sizes': [[2, 2], [2, 2], [2, 2], [2, 2], [2, 2]], 'conv_kernel_sizes': [[3, 3], [3, 3], [3, 3], [3, 3], [3, 3], [3, 3]], 'do_dummy_2D_data_aug': False}}
_

**Comment by ysssgdhr** [comment]
update：
The task is from OURS to KITS, from difficult to easy. I put the data of KITS and OURS together, and the label of KITS is empty. Then modify a small part of the code to allow plan_and_preprocess to run. I follow the final generated plans.pkl file to use my own unsupervised method for training and inference.
But I found a fatal problem. For OURS, the data is isotropic, so the prediction direction does not matter. But for kits, the model is not predicted or trained from the z-axis (I have not touched the plans.pkl file).
![image](https://user-images.githubusercontent.com/24486813/100340854-c2c47200-3016-11eb-8219-f50d4623720f.png)

I think I can choose one of the following options to solve the problem:
1. Forget the OURS data set, let nnunet generate plans.pkl file based on the img and gt of the kits data set and use it for unsupervised domain migration training.
2. Directly modify the 'transpose_forward' and 'transpose_backward' in the plans.pkl file.
3. Exchange the axis of the '*.nii.gz' file. (I doubt this will be useful, because nnunet is likely to change the axis back)
I personally think that the first one seems feasible. I uploaded the sample ct file and plans file. It would be very nice if there is any suggestion. :-)

[kits_case_sample.nii.gz](https://github.com/MIC-DKFZ/nnUNet/files/5602296/kits_case_sample.nii.gz)
[ours_case_sample.nii.gz](https://github.com/MIC-DKFZ/nnUNet/files/5602311/ours_case_sample.nii.gz)
[model_final_checkpoint.txt](https://github.com/MIC-DKFZ/nnUNet/files/5602339/model_final_checkpoint.txt)
rename 'model_final_checkpoint.txt''s suffix to '.pkl' please, there are some weird restrictions on uploaded file format.

**Comment by FabianIsensee** [comment]
Hi,
sorry I am extremely short on time right now. I will read through this more carefully next week.
I have downloaded the example data you provided and run the 3d_fullres KiTS model. My output looks good:
![image](https://user-images.githubusercontent.com/11302164/100342981-0e5b2980-2fdf-11eb-9479-325681fdd2c7.png)

Can you provide cases where the KiTS model fails so that I can reproduce your problem?

What  I did is the following:

`nnUNet_download_pretrained_model Task048_KiTS_clean`
(I did not have the model on my workstation at home, so I downloaded the pretrained one)

`nnUNet_predict -t 48 -i tmp -o tmp_out`

I predicted both the KiTS and the OURS example. Both worked fine.

Best,
Fabian

**Comment by ysssgdhr** [comment]
Thank you for your attention. I fully understand your limited time and my question is not urgent. You can wait until you have time to answer. :)
But my project is to improve the model training on OURS inference on KITS. I'II simply my questions into several below.
1、Evaluation metrics during training can hit up to 70%. My validation dataset IS THE test dataset. But after inference, DICE is only 23% for tumor. I think evaluation metrics is the average dice of kidney and tumor without background. Maybe I am wrong?
2、You said there is no reason to use batchsize more than 2 when using 3DUnet. Here I am using 2DUnet, and nnUnet suggestted batchsize is 104. Is this normal?

Thank you again for your attention. My question is not urgent , so take your time. :)

**Comment by FabianIsensee** [comment]
I will have a lot more time next week, but let me answer quickly here:
1) the evaluation metric during training is not real. It's an optimistic estimate and it is very well possible that the actual metric is much smaller, especially if you have images where no tumor is present
2) I don't remember saying that this is true in general. There are situations in which you want more batch size. nnU-Net is implemented such that it will increase the batch size if there is gpu memory headroom available. Most of the time increasing the batch size beyond 2 in 3D U-Nets only increases your training time and does not meaningfully improves the segmentation. If you notice that your model has trouble fitting the data (no over-fitting) then increasing the batch size may be beneficial. Having a large batch size for the 2D model is completely normal.
Best,
Fabian
